Reference assembly generation

mrward opened the pull request:
Support generating reference assemblies for PCL profiles.

Add a ReferenceAssemblyFramework item to a NuGet packaging project (.nuproj):

```
<ItemGroup>
    <ReferenceAssemblyFramework Include=".NETPortable,Version=v4.5,Profile=Profile259" />
</ItemGroup>
```

Then building the .nuproj will run the ApiIntersect tool against the referenced projects and generate a reference assembly that is included in the NuGet package.

mrward committed:
Add ApiIntersect submodule

mrward committed:
Updated reference to ApiIntersect

Added binding redirect which fixes the FileLoadException for
Mono.Cecil on Windows.

mrward committed:
Support generating reference assemblies

Adding a ReferenceAssemblyFramework item to a NuGet packaging project
will now generate reference assemblies for the specified Portable
Class Library profile for the referenced projects.

<ReferenceAssemblyFramework Include=".NETPortable,Version=v4.5,Profile=Profile259" />

mrward committed:
Generate reference assemblies for all referenced projects

All referenced projects that will be included in the NuGet package
will be used when generating reference assemblies. Previously only
those which had IsPackable=true were included.

mrward committed:
Add reference assembly tests

Checks that the GetPackageContents target includes the reference
assembly.

mrward committed:
Fix incremental build for reference assemblies.

Prevent the reference assemblies being built every time when the
input assemblies being intersected have not changed.

mrward committed:
Provide a way to get ApiIntersect output paths without building

Can now call GetPackageContents target without having to build the
project in order for reference assemblies to be included. Previously
the GetPackageContents target would run the ApiIntersect tool which
would build the reference assemblies which could fail since other
project outputs may not have been built.

mrward committed:
Add tests for GetApiIntersectTargetPaths task.

kzu commented on src/Build/NuGet.Build.Packaging.Tasks/ApiIntersect.cs:
```diff
@@ -0,0 +1,106 @@
+				return GetPortableLibraryPath(framework);
+			}
+
+			throw new NotImplementedException("Only PCL profiles supported.");
```
what about `netstandard`? not worth it for now?

kzu commented on src/Build/NuGet.Build.Packaging.Tasks/ApiIntersect.cs:
```diff
@@ -0,0 +1,106 @@
+
+			return Path.Combine(
+				Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86, Environment.SpecialFolderOption.DoNotVerify),
+				@"Reference Assemblies\Microsoft\Framework\.NETPortable");
```
@victorgarciaaprea do we know if this will work at all in `dev15`? What's the guideline for referring to the applocal MSBuild there?

@mrward the challenge here is that starting in dev15, VS will install everything under the VS install directory, allowing for true SxS installs, so we need to make sure this doesn't break there.

kzu commented on src/Build/NuGet.Build.Packaging.Tasks/GenerateReferenceAssembly.csproj:
```diff
@@ -0,0 +1,44 @@
+      <_CompileFiles>$([System.IO.Directory]::GetFiles('$(SrcDir)Contract\', '*.cs', SearchOption.AllDirectories))</_CompileFiles>
+    </PropertyGroup>
+    <ItemGroup>
+      <Compile Include="$(_CompileFiles)" />
```
Can't we just use `Include="$(SrcDir)Contract\**\*.*" />`?

mrward replied:
Yes that will most likely work. I used a workaround for xbuild but since we are not supporting xbuild I can remove it. I will make sure it works on the Mac.

mrward replied:
Updated this to use a wildcard.

mrward committed:
Simplify file collection when generating reference assemblies

Switch from using a property function to a wildcard when finding
files to compile for the reference assembly.

kzu reviewed (approved):
👍
